File: LeetCode/1916. Count Ways to Build Rooms in an Ant Colony/solution.cpp
// class Solution {
// private:
//     static constexpr int mod = 1000000007;
//     using LL = long long;

// public:
//     int waysToBuildRooms(vector<int>& prevRoom) {
        
//         // 快速幂计算 x^y
//         auto quickmul = [&](int x, int y) {
//             int ret = 1, cur = x;
//             while (y) {
//                 if (y & 1) {
//                     ret = (LL)ret * cur % mod;
//                 }
//                 cur = (LL)cur * cur % mod;
//                 y >>= 1;
//             }
//             return ret;
//         };
        
//         int n = prevRoom.size();
//         // fac[i] 表示 i!
//         // inv[i] 表示 i! 的乘法逆元
//         vector<int> fac(n), inv(n);
//         fac[0] = inv[0] = 1;
//         for (int i = 1; i < n; ++i) {
//             fac[i] = (LL)fac[i - 1] * i % mod;
//             // 使用费马小定理计算乘法逆元
//             inv[i] = quickmul(fac[i], mod - 2);
//         }
        
//         // 构造树
//         vector<vector<int>> edges(n);
//         for (int i = 1; i < n; ++i) {
//             edges[prevRoom[i]].push_back(i);
//         }
        
//         vector<int> f(n), cnt(n);
//         function<void(int)> dfs = [&](int u) {
//             f[u] = 1;
//             for (int v: edges[u]) {
//                 dfs(v);
//                 // 乘以左侧的 f[ch] 以及右侧分母中 cnt[ch] 的乘法逆元
//                 f[u] = (LL)f[u] * f[v] % mod * inv[cnt[v]] % mod;
//                 cnt[u] += cnt[v];
//             }
//             // 乘以右侧分子中的 (cnt[i] - 1)!
//             f[u] = (LL)f[u] * fac[cnt[u]] % mod;
//             ++cnt[u];
//         };
        
//         dfs(0);
//         return f[0];
//     }
// };


// // 作者：LeetCode-Solution
// // 链接：https://leetcode-cn.com/problems/count-ways-to-build-rooms-in-an-ant-colony/solution/tong-ji-wei-yi-qun-gou-zhu-fang-jian-de-uqhn7/
// // 来源：力扣（LeetCode）
// // 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。


using ll = long long;
const ll MOD = 1e9 + 7;
const int N = 100005;

bool inited = false;
ll fac[N], ifac[N], ans[N];
int sz[N];
vector<vector<int>> adj;
ll fexp(ll x, ll y) {
    ll ans = 1;
    while (y) {
        if (y & 1)
            ans = ans * x % MOD;
        x = x * x % MOD;
        y >>= 1;
    }
    return ans;
}
ll C(int n, int k) {
    return fac[n] * ifac[k] % MOD * ifac[n - k] % MOD;
}

void init() {
    inited = true;
    fac[0] = ifac[0] = 1;
    for (int i = 1; i < N; ++i)
        fac[i] = fac[i - 1] * i % MOD;
    ifac[N - 1] = fexp(fac[N - 1], MOD - 2);
    for (int i = N - 2; i >= 1; --i)
        ifac[i] = ifac[i + 1] * (i + 1) % MOD;
}

void dfs(int u) {
    sz[u] = 0;
    for (int v : adj[u]) {
        dfs(v);
        sz[u] += sz[v];
    }
    
    int tot = sz[u];
    ans[u] = 1;
    for (int v : adj[u]) {
        ans[u] = ans[u] * ans[v] % MOD;
        ans[u] = ans[u] * C(tot, sz[v]) % MOD;
        tot -= sz[v];
    }
    
    sz[u]++;
}

class Solution {
public:
    int waysToBuildRooms(vector<int>& prevRoom) {
        if (!inited)
            init();
        
        int n = prevRoom.size();
        adj = vector<vector<int>>(n);
        for (int i = 1; i < n; ++i)
            adj[prevRoom[i]].emplace_back(i);
        
        dfs(0);
        
        return ans[0];
    }
};


// 作者：吴自华
// 链接：https://leetcode-cn.com/circle/discuss/6LY3tn/view/TG6U7W/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: LeetCode/1916. Count Ways to Build Rooms in an Ant Colony/solution.py
def quick_mul(a, p):
    ans = 1
    while p:
        if p&1:
            ans *= a 
            ans %= MOD 
        a *= a 
        a %= MOD 
        p = (p>>1)
    return ans 

n = 10**5+5
MOD = 10**9+7

factors = [1]*n 
rev_factors = [1]*n
for i in range(1,n):
    factors[i] = i*factors[i-1]%MOD
# for i in range(1,n):
#     rev_factors[i] = quick_mul(factors[i], MOD-2)%MOD 

rev_factors[n-1] = pow(factors[n-1], MOD-2, MOD) 
for i in range(1,n-1)[::-1]:
    # rev_factors[i] = quick_mul(factors[i], MOD-2)%MOD 
    rev_factors[i] = rev_factors[i+1]*(i+1)%MOD

class Solution:
    def waysToBuildRooms(self, prevRoom: List[int]) -> int:
        
        graph = defaultdict(list)
        for cur, pre in enumerate(prevRoom):
            if pre >= 0:
                graph[pre].append(cur)

        

      
        def dfs(room):
            res, tot = 1 , 0
            for nxt in graph[room]:
                path, sz = dfs(nxt)
                tot += sz 
                res *= path
                res *= rev_factors[sz]
                res %= MOD
            res *= factors[tot]
            res %= MOD
            # print(fac, tot, tot2)
            
            return res, tot+1

        return dfs(0)[0]




class Solution:
    def waysToBuildRooms(self, arr: int) -> int:
        MOD = 10**9 + 7
        g = collections.defaultdict(list)
        for cur, pre in enumerate(arr):
            g[pre].append(cur)
            
        def dfs(cur):
            if not g[cur]:
                return 1, 1
            ans, l = 1, 0
            for nxt in g[cur]:
                tmp, r = dfs(nxt)
                ans = (ans * tmp * math.comb(l+r, r)) % MOD
                l += r
            return ans, l + 1
            
        return dfs(0)[0]



from math import comb
from functools import reduce


class Solution:
    def waysToBuildRooms(self, prevRoom: List[int]) -> int:
        MOD = 10**9+7
        graph = defaultdict(list)
        for cur, pre in enumerate(prevRoom):
            if pre >= 0:
                graph[pre].append(cur)
      
        def dfs(room):
            res = 1 
            child = []
            for nxt in graph[room]:
                path, sz = dfs(nxt)
                child.append(sz)
                res *= path
                res %= MOD
            tot2 = tot = sum(child)
            for sz in child:
                res *= comb(tot,sz)
                res %= MOD
                tot -= sz
            # print(fac, tot, tot2)
            
            return res, tot2+1

        return dfs(0)[0]

